### PR TITLE
Add rule for detecting manual reimplementation of 'collections' funtionality

### PR DESCRIPTION
**Pull request by mschwager**
Detects manual implementation of [`collections.Counter`](https://docs.python.org/3.8/library/collections.html#collections.Counter) and [`collections.defaultdict`](https://docs.python.org/3.8/library/collections.html#collections.defaultdict).

Tested with:

```
$ docker run --rm -v $(pwd):/home/repo returntocorp/sgrep:develop -f /home/repo/python/smells/manual-collections-create.yaml /home/repo/python/smells/manual-collections-create.py | grep "manually creating a"
running 4 rules from 1 yaml files (0 yaml files were invalid)
rule:python.smells.manual-counter-create: manually creating a counter - use collections.Counter
rule:python.smells.manual-defaultdict-dict-create: manually creating a defaultdict - use collections.defaultdict(dict)
rule:python.smells.manual-defaultdict-dict-create: manually creating a defaultdict - use collections.defaultdict(dict)
rule:python.smells.manual-defaultdict-list-create: manually creating a defaultdict - use collections.defaultdict(list)
rule:python.smells.manual-defaultdict-list-create: manually creating a defaultdict - use collections.defaultdict(list)
rule:python.smells.manual-defaultdict-set-create: manually creating a defaultdict - use collections.defaultdict(set)
rule:python.smells.manual-defaultdict-set-create: manually creating a defaultdict - use collections.defaultdict(set)
```

### CODE REVIEW

File: python/smells/manual-collections-create.yaml [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        $DICT = {}
+        ...
+        for $KEY, $VALUE in $OTHERDICT:
+            ...
+            if $KEY not in $DICT:
+                ...
+                $DICT[$KEY] = 0
+                ...
+            $DICT[$KEY] += 1
+    message: "manually creating a counter - use collections.Counter"
```
Comment (ievans) [inline python/smells/manual-collections-create.yaml]: I did not know about collections.Counteer -- TIL!

File: python/smells/manual-collections-create.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+foo = {}
+
+# ruleid:manual-defaultdict-dict-create
```
Comment (ievans) [inline python/smells/manual-collections-create.py]: this is great, can you add just a few negative test cases though?

Comment (mschwager) [reply python/smells/manual-collections-create.py]: Yup, added a few!

**Comment by mschwager** [comment]
Looks like unrelated tests are failing:

```
 failing rule files: 
 ✖ FAILED rule file: python/deadcode/baseclass-attribute-override.yaml check: baseclass-attribute-override
              in test: /github/workspace/python/deadcode/baseclass-attribute-override.py, expected lines: set() != reported: {2}
1 checks failed tests (run with verbose flag for more details)
```

**Comment by mschwager** [comment]
Running this rule over a large corpus didn't produce any results: https://dev.massive.ret2.co/triager/filter/600

Not sure if the rule is faulty, or if this is just an uncommon pattern. Either way, this likely won't make it into the "golden rules."

**Pull request merged**